Adds docs for creating custom control panel (#1909)
plone/documentation: docs/developer-guide/create-control-panel.md

These values correspond to the groups in Site Setup.

`plone-general`
:   General settings

`plone-content`
:   Content-related settings

`plone-users`
:   Users and groups settings

`plone-security`
:   Security settings

`plone-advanced`
:   Advanced settings

### Set default values in the registry

Define default values for your settings in {file}`mypackage/profiles/default/registry.xml`.

```xml
<!-- mypackage/profiles/default/registry.xml -->
<?xml version="1.0"?>
<registry>
    <records interface="mypackage.controlpanel.settings.IMyControlPanelSettings"
             prefix="my.addon">
        <value key="my_setting">default value</value>
        <value key="my_choice">value3</value>
    </records>
</registry>
```

### Access your settings in code

You can access your settings in Python code as follows.

```python
from plone.registry.interfaces import IRegistry
from zope.component import getUtility

registry = getUtility(IRegistry)
settings = registry.forInterface(IMyControlPanelSettings, prefix="my.addon")

# Now you can access the settings
my_setting_value = settings.my_setting
my_choice_value = settings.my_choice
```

> ## Register a control panel

Reviewer: I got a little bit lost here. Does this heading apply to both the plonecli and manual options, or just the manual one?

- If both, then we should add a sentence to clarify this.
- If only the latter, then its heading level should increase.

Reviewer: Does the Plone site need to be restarted after adding all the files for the control panel to show up? If so, then that step also needs to be documented in the appropriate location.

Author: > Does this heading apply to both the plonecli and manual options, or just the manual one?

plonecli registers itself but manually we need to register for the `manual` steps only

> Does the Plone site need to be restarted after adding all the files for the control panel to show up? If so, then that step also needs to be documented in the appropriate location.

Yes, a restart is needed

Reviewer: ```suggestion
### Register a control panel
```

    </records>
</registry>
```

### Access your settings in code

You can access your settings in Python code as follows.

```python
from plone.registry.interfaces import IRegistry
from zope.component import getUtility

registry = getUtility(IRegistry)
settings = registry.forInterface(IMyControlPanelSettings, prefix="my.addon")

# Now you can access the settings
my_setting_value = settings.my_setting
my_choice_value = settings.my_choice
```

## Register a control panel

To manually register a view as a control panel, add the following registration to your {file}`/profiles/default/controlpanel.xml`.

```xml
<?xml version="1.0"?>
  <object
      name="portal_control-panel"
      xmlns:i18n="http://xml.zope.org/namespaces/i18n"
      i18n:domain="lmu.behavior">
    <configlet
      title="Some Control Panel"
      action_id="collective.example.some_control-panel"
      appId="collective.example"
      category="Products"
      condition_expr=""
      url_expr="string:${portal_url}/@@some_view"
      icon_expr=""
      visible="True"
      i18n:attributes="title">
          <permission>Manage portal</permission>
    </configlet>
  </object>
```

> 

Reviewer: ````suggestion

After you perform the above steps for the manual process, you must restart the Plone site with the following command.

```shell
@rohnsha0 to provide command here
```

````

Author: just a simple restart will do! no special commands reqd.

Reviewer: How will a first-timer know that command? Please provide it.

Author: just closing the running server and then restarting again with `./bin/instange fg` for running with buildouts?

something like above?! would be greate? @stevepiercy

Reviewer: I think you now realize the problem I was trying to convey. Cookieplone, pip, and buildout installations might have different procedures. To stop an instance, I believe it is always:

```markdown
{kbd}`ctrl-c`
```

To start:

- [buildout](https://6.docs.plone.org/admin-guide/install-buildout.html#start-plone-in-foreground-mode): `bin/instance fg`
- [pip](https://6.docs.plone.org/admin-guide/install-pip.html#start-plone-in-foreground-mode): `bin/runwsgi -v instance/etc/zope.ini`
- [Cookieplone (backend)](https://6.docs.plone.org/install/create-project-cookieplone.html#start-plone-backend): `make backend-start`

...and if you have Volto, then:

- [Cookieplone (frontend)](https://6.docs.plone.org/install/create-project-cookieplone.html#start-plone-frontend): `make backend-start` in one shell session window, and `make frontend-start` in another

We can use the markup for tabs from Sphinx Design for this.

https://6.docs.plone.org/contributing/documentation/myst-reference.html#tabs

Author: @stevepiercy the tabs do not support code block ig... can u recheck if u can add code blocks inside a tab

```
Traceback
=========

      File "/home/rohnsha0/code/plone/documentation/venv/lib/python3.13/site-packages/sphinx_design/tabs.py", line 244, in run
        tab_label, tab_content = tab_item.children
        ^^^^^^^^^^^^^^^^^^^^^^
    ValueError: not enough values to unpack (expected 2, got 1)

```

Reviewer: Strange, I got a notification that you were struggling with the tabs, but I don't see that comment in the pull request or code sample of what you tried, @rohnsha0.

Author: done! adding instructions

Author: > Strange, I got a notification that you were struggling with the tabs, but I don't see that comment in the pull request or code sample of what you tried, @rohnsha0.

hahaaa, that was me getting frustrated due to errors while adding code block inside tabs... however later decided to it by myself only!

Reviewer: See my revisions in https://github.com/plone/documentation/pull/1909/commits/2a65a7948c995bbc69049a191afbc746c365338b and reasoning in the commit message.

> ## Use `FieldSet` to group fields

Reviewer: Is this also part of the manual process only or both manual and plonecli?

Author: @stevepiercy this needs to be added in any way you use to create the controlpanel, as these include the fields (to be grouped) that needs to be shown in the settings (of the controlpanel)

Reviewer: OK, please see other comments regarding the flow and structure of this chapter. I think we're close to finished with those enhancements.

:   For selection from a list of values

`schema.Datetime`
:   For date and time values

`schema.List`
:   For list of values

## Modify control panel fields

When you modify the fields in your control panel settings interface, the changes won't be automatically reflected in existing sites.
You'll need to perform one or more of the following steps.

-   Run the appropriate upgrade steps.
-   Reinstall your add-on.
-   Test with a fresh site installation.

## Troubleshooting

If your control panel doesn't appear or doesn't work as expected:

-   Verify that all ZCML is properly registered
-   Check for errors in the Plone error log
-   Ensure your GenericSetup profiles are correctly installed
-   Validate that the interface path in registry.xml matches your actual Python path

## Example file structure

Below is a complete example file structure for a basic add-on with a control panel.

```
mypackage/
├── __init__.py
├── configure.zcml
├── controlpanel/
│   ├── __init__.py
│   ├── configure.zcml
│   └── settings.py
└── profiles/
    └── default/
        ├── controlpanel.xml
        ├── metadata.xml
        └── registry.xml
```

## REST API compatible control panels

> For better integration between backend and Volto, you can create REST API compatible control panels using the adapter pattern. This approach is particularly useful when developing control panels that need to work seamlessly with Volto.

Reviewer: One sentence per line.

```suggestion
For better integration between backend and Volto, you can create REST API compatible control panels using the adapter pattern.
This approach is particularly useful when developing control panels that need to work seamlessly with Volto.
```

> Create a Python module like `mypackage/controlpanel.py`:

Reviewer: ```suggestion
Create a Python module {file}`mypackage/controlpanel.py` as follows.
```

```python
from plone.restapi.controlpanels import RegistryConfigletPanel
from zope.component import adapter
from zope.interface import Interface
from zope.i18nmessageid import MessageFactory

_ = MessageFactory("mypackage")

@adapter(Interface, Interface)
class MyAddonControlPanel(RegistryConfigletPanel):

>     """Volto-compatible REST API control panel for my addon settings."""

Reviewer: ```suggestion
    """Volto-compatible REST API control panel for my add-on settings."""
```

    schema = IMyControlPanelSettings
    schema_prefix = "my.addon"
    configlet_id = "my-controlpanel"
    configlet_category_id = "plone-general"
    title = _("My Addon Settings")
    group = "General"
```

> Then register the adapter in your ZCML configuration file:

Reviewer: ```suggestion
Then register the adapter in your ZCML configuration file.
```

```xml
<!-- mypackage/configure.zcml or mypackage/controlpanel/configure.zcml -->
<configure
    xmlns="http://namespaces.zope.org/zope"
    i18n_domain="mypackage">

    <adapter
        factory=".controlpanel.MyAddonControlPanel"
        name="my-controlpanel"
    />

</configure>
```

> The `group` property in the control panel class corresponds to the control panel category in Volto:
> - `General`: General settings (corresponds to `plone-general`)
> - `Content`: Content-related settings (corresponds to `plone-content`)
> - `Users`: Users and groups settings (corresponds to `plone-users`)
> - `Security`: Security settings (corresponds to `plone-security`)
> - `Advanced`: Advanced settings (corresponds to `plone-advanced`)

Reviewer: Convert to definition list.

```suggestion
The `group` property in the control panel class corresponds to the control panel category in Volto.

`General`
:   General settings (corresponds to `plone-general`)

`Content`
:   Content-related settings (corresponds to `plone-content`)

`Users`
:   Users and groups settings (corresponds to `plone-users`)

`Security`
:   Security settings (corresponds to `plone-security`)

`Advanced`
:   Advanced settings (corresponds to `plone-advanced`)
```

With this approach, your control panel will be automatically available through the REST API at the endpoint `@controlpanels/my-controlpanel`, making it easy to integrate with Volto without additional configuration.

> You will still need to set up the registry.xml with default values as described earlier.

Reviewer: ```suggestion
You will still need to set up {file}`registry.xml` with default values as described earlier.
```

---
myst:
  html_meta:
    "description": "How to create a control panel"
    "property=og:description": "How to create a control panel"

>     "property=og:title": "Create a control panel"

Reviewer: ```suggestion
    "description": "How to create a control panel in Plone for Classic UI and Volto"
    "property=og:description": "How to create a control panel in Plone for Classic UI and Volto"
    "property=og:title": "Create a control panel in Plone for Classic UI and Volto"
```

>     "keywords": "Plone, create, control panel, plonecli, registry"

Reviewer: ```suggestion
    "keywords": "Plone, create, control panel, plonecli, registry, Classic UI, Volto, frontend, backend"
```

---

(backend-controlpanels-label)=

# Create a control panel

> There are two approaches to create a control panel for your Plone add-on:

Reviewer: After reading this over, I found its flow disorienting. This suggestion should help.

Additionally, the two subsections and their subheadings should each go one level deeper with one more `#`.

```suggestion
This chapter describes how to create a control panel for your Plone add-on, whether accessed through either the Classic UI or Volto frontend.

It also covers advanced topics—including how to group fields in your control panel—and provides a schema field reference, troubleshooting tips, control panel file structure, and a Plone REST API compatibility reference.

## Creation approaches

There are two approaches to create a control panel for your Plone add-on:
```

-   [`plonecli`](https://pypi.org/project/plonecli/)
-   manual

## `plonecli`

To add a control panel to your add-on, you can use [`plonecli`](https://pypi.org/project/plonecli/) as follows.

```shell
plonecli add controlpanel
```

> This creates the control panel Python file in the control panel's folder where you can define your control panel schema fields.

Reviewer: Where should the reader go from here? I'm lost. Help! In other words, write documentation as if you didn't have years of experience with Plone.

Author: Okay, after running the `plonecli` command to create controlpanel. It creates control panel with a basic field (schema.TextLine) for the CLassic and Volto both, like an angle 😇😎

So, ig the reader will now be required to know the schemas, then `Fieldset` if needed to group fields together

Reviewer: In that case, we need to direct the reader to those parts of the documentation.

And now that I think about it, we should have only one method. plonecli automatically performs all the steps in the manual process for the reader, so it's silly to say that there are two processes. The steps are identical, and the only difference is automation of performing them. In any case, the reader will need to modify their add-on's schema, as plonecli has no way to read the developer's mind. Yet.

Author: IMO, we can simply add a note that plonecli does everything done in the manual process... since, the manual process exposes developers to some minor technicalities of the thing

Make sure to include the above file in your package's main {file}`mypackage/configure.zcml` as shown by the highlighted line below.

{emphasize-lines="9"}
```xml
<!-- mypackage/configure.zcml -->
<configure
    xmlns="http://namespaces.zope.org/zope"
    xmlns:i18n="http://namespaces.zope.org/i18n"
    i18n_domain="mypackage">
    
    <!-- Other configuration -->
    
    <include package=".controlpanel" />
    
</configure>
```

### Add the control panel entry

Create a {file}`mypackage/profiles/default/controlpanel.xml` in your package's GenericSetup profile with the following content to add your control panel to the Plone control panel listing.

```xml
<!-- mypackage/profiles/default/controlpanel.xml -->
<?xml version="1.0"?>
<object name="portal_controlpanel">
    <configlet
        title="My Addon Settings"
        action_id="my-controlpanel"
        appId="my.addon"
        category="plone-general"
        condition_expr=""
        icon_expr="string:puzzle"
        url_expr="string:${portal_url}/@@my-controlpanel"
        visible="True">
        <permission>Manage portal</permission>
    </configlet>
</object>
```

The category attribute can be one of the following values.

> These values correspond to the groups in Site Setup.

Reviewer: ```suggestion
These values correspond to the groups in {guilabel}`Site Setup`.
```

`plone-general`
:   General settings

`plone-content`
:   Content-related settings

`plone-users`
:   Users and groups settings

`plone-security`
:   Security settings

`plone-advanced`

> :   Advanced settings

Reviewer: ```suggestion
`plone-general`
:   {guilabel}`General`

`plone-content`
:   {guilabel}`Content`

`plone-users`
:   {guilabel}`Users`

`plone-security`
:   {guilabel}`Security`

`plone-advanced`
:   {guilabel}`Advanced`
```

    </records>
</registry>
```

### Access your settings in code

You can access your settings in Python code as follows.

```python
from plone.registry.interfaces import IRegistry
from zope.component import getUtility

registry = getUtility(IRegistry)
settings = registry.forInterface(IMyControlPanelSettings, prefix="my.addon")

# Now you can access the settings
my_setting_value = settings.my_setting
my_choice_value = settings.my_choice
```

### Register a control panel

To manually register a view as a control panel, add the following registration to your {file}`/profiles/default/controlpanel.xml`.

```xml
<?xml version="1.0"?>
  <object
      name="portal_control-panel"
      xmlns:i18n="http://xml.zope.org/namespaces/i18n"
      i18n:domain="lmu.behavior">
    <configlet
      title="Some Control Panel"
      action_id="collective.example.some_control-panel"
      appId="collective.example"
      category="Products"
      condition_expr=""
      url_expr="string:${portal_url}/@@some_view"
      icon_expr=""
      visible="True"
      i18n:attributes="title">
          <permission>Manage portal</permission>
    </configlet>
  </object>
```

> 
> ## Use `FieldSet` to group fields

Reviewer: This needs to transition the reader more clearly from the manual process. The page introduction also needs some enhancement to reflect the page structure.

```suggestion
Your control panel should now appear in {guilabel}`Site Setup`.

## Use `FieldSet` to group fields
```

Author: ah, agreed!.

        description=u'Enter the value for my setting',
        required=False
    )
    
    # Advanced settings
    advanced_setting1 = schema.TextLine(
        title=u'Advanced Setting 1',
        required=False
    )
    
    advanced_setting2 = schema.Bool(
        title=u'Advanced Setting 2',
        default=False
    )
```

## Common schema fields

The following is a list of commonly used schema field types.

`schema.TextLine`
:   For single-line text

`schema.Text`
:   For multi-line text

`schema.Bool`
:   For boolean values

`schema.Int`
:   For integer values

`schema.Float`
:   For floating-point values

`schema.Choice`
:   For selection from a list of values

`schema.Datetime`
:   For date and time values

`schema.List`
:   For list of values

## Modify control panel fields

When you modify the fields in your control panel settings interface, the changes won't be automatically reflected in existing sites.
You'll need to perform one or more of the following steps.

-   Run the appropriate upgrade steps.
-   Reinstall your add-on.
-   Test with a fresh site installation.

## Troubleshooting

If your control panel doesn't appear or doesn't work as expected:

> -   Verify that all ZCML is properly registered
> -   Check for errors in the Plone error log
> -   Ensure your GenericSetup profiles are correctly installed
> -   Validate that the interface path in registry.xml matches your actual Python path

Reviewer: How does a first-timer do any of the following? I'm a first-timer. I have no clue. Should I go back to a specific step or another part of the documentation? Help!

```suggestion
-   Verify that all ZCML is properly registered.
-   Check for errors in the Plone error log.
-   Ensure your GenericSetup profiles are correctly installed.
-   Validate that the interface path in {file}`registry.xml` matches your actual Python path.
```

Author: Umm, the complexity of Plone. 
If using plonecli, it just works. Issues are most probable in performing the steps manually especially the registering things in `configure.zcml`. So registering things might be have to be taken a look at.
but in my experience, the control panel if not appearing then either due to not registering components properly
or due to some errors which appears on the server logs

Reviewer: Huh? My question is not whether someone needs to perform these troubleshooting tasks. My question is, how does a first-timer perform each of them? We need to write documentation as if we don't have years of experience with Plone, and think like a first-timer.

## Example file structure

Below is a complete example file structure for a basic add-on with a control panel.

```
mypackage/
├── __init__.py
├── configure.zcml
├── controlpanel/
│   ├── __init__.py
│   ├── configure.zcml
│   └── settings.py
└── profiles/
    └── default/
        ├── controlpanel.xml
        ├── metadata.xml
        └── registry.xml
```

> ## REST API compatible control panels

Reviewer: ```suggestion
## REST API compatibility
```

> For better integration between backend and Volto, you can create REST API compatible control panels using the adapter pattern.

Reviewer: ```suggestion
For better integration between Plone's backend and its frontend Volto, you can create REST API compatible control panels using the adapter pattern.
```
